client/network/src/discovery.rs

```diff
@@ -144,48 +141,71 @@ impl DiscoveryConfig {
 
 	/// Add discovery via Kademlia for the given protocol.
 	pub fn add_protocol(&mut self, id: ProtocolId) -> &mut Self {
-		let name = protocol_name_from_protocol_id(&id);
-		self.add_kademlia(id, name);
```

Comment:
Creation of `Kademlia` instances is moved into `fn finish`. Otherwise users could first call `add_protocol` and then `use_kademlia_disjoint_query_paths` with the latter not taking any effect given that the `Kademlia`s are already created.